firefox_ui_harness/testcases/update.py

```diff
@@ -0,0 +1,133 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+from ..testcase import FirefoxTestCase
+
+
+class UpdateTestCase(FirefoxTestCase):
+
+    def __init__(self, *args, **kwargs):
+        FirefoxTestCase.__init__(self, *args, **kwargs)
+
+        self.target_buildid = kwargs.pop('update_target_buildid')
+        self.target_version = kwargs.pop('update_target_version')
+
+        self.update_channel = kwargs.pop('update_channel')
+        self.update_mar_channels = set(kwargs.pop('update_mar_channels'))
+
+        self.updates = []
+
+    def setUp(self, is_fallback=False):
+        FirefoxTestCase.setUp(self)
+
+        self.current_update_index = 0
+        self.staging_directory = self.software_update.staging_directory
+
+        # If requested modify the default update channel. It will be active
+        # after the next restart of the application
+        # TODO: Bug 1142805 - Modify file via Python directly
+        if self.update_channel:
+            # Backup the original content and the path of the channel-prefs.js file
+            self.default_update_channel = {
+                'content': self.software_update.update_channel.file_contents,
+                'path': self.software_update.update_channel.file_path,
+            }
+            self.software_update.update_channel.default_channel = self.update_channel
+
+        # If requested modify the list of allowed MAR channels
+        # TODO: Bug 1142805 - Modify file via Python directly
+        if self.update_mar_channels:
+            # Backup the original content and the path of the update-settings.ini file
+            self.default_mar_channels = {
+                'content': self.software_update.mar_channels.config_file_contents,
+                'path': self.software_update.mar_channels.config_file_path,
+            }
+            self.software_update.mar_channels.add_channels(self.update_mar_channels)
+
+        # Bug 1142805 - Until we don't modify the channel-prefs.js and update-settings.ini
+        # files before Firefox gets started, a restart of Firefox is necessary to
+        # accept the new update channel.
+        self.restart()
+
+        # Dictionary which holds the information for each update
+        self.updates = [{
+            'build_pre': self.software_update.build_info,
+            'build_post': self.software_update.build_info,
```

Comment:
We have to do this to workaround the fact that a possible crash or whatever else doesn't let us finish the test. If we move this assignment to the end of the test, we won't have a post build assigned. This would add a lot more extra cases to the dashboard (treeherder view).

Even that we are safer now for Python tests, I would propose that we keep this for now and if necessary find an alternative solution as a follow-up bug. The code as is has been proven to work over the last years.